File: 2015/12/12.go
package main

import "fmt"
import "strconv"
import "io/ioutil"
import "strings"

func sumInFile(file string) int {
	s := 0
	data, _ := ioutil.ReadFile(file)
	for _, line := range strings.Split(string(data), "\n") {
		s += numberSum(line)
	}
	return s
}

func numberSum(s string) int {
	sum := 0
	n := false
	curr := ""
	for i, c := range s {
		chr := string(c)
		switch chr {
		case "0", "1", "2", "3", "4", "5", "6", "7", "8", "9":
			if !n {
				if string(s[i-1]) == "-" {
					curr = "-"
				}
			}
			curr += chr
			n = true
		default:
			if n {
				num, _ := strconv.Atoi(curr)
				sum += num
				n = false
				curr = ""
			}
		}
	}
	return sum
}

func main() {
	fmt.Println("Part 1:", sumInFile("input"))
}

File: 2015/12/12.py
import json

data = []
with open("input") as f:
    for line in f.readlines():
        data.append(json.loads(line))

def recursive_sum(obj, s=0):
    if isinstance(obj, list):
        s += sum(recursive_sum(o) for o in obj)
    if isinstance(obj, dict):
        s += sum(recursive_sum(o) for o in obj.values())
    if isinstance(obj, int):
        s += obj
    return s

def recursive_red_sum(obj, s=0):
    if isinstance(obj, list):
        s += sum(recursive_red_sum(o) for o in obj)
    if isinstance(obj, dict):
        if 'red' not in obj.values():
            s += sum(recursive_red_sum(o) for o in obj.values())
    if isinstance(obj, int):
        s += obj
    return s

print("Part 1:", recursive_sum(data))
print("Part 2:", recursive_red_sum(data))


